### PR TITLE
Fix #383: Implement CA1034: NestedTypesShouldNotBeVisible

### LINKED ISSUE #383: Port FxCop rule CA1034: NestedTypesShouldNotBeVisible
**Title:** Nested types should not be visible

**Description:**

A nested type is a type that is declared in the scope of another type. Nested types are useful to encapsulate private implementation details of the containing type. Used for this purpose, nested types should not be externally visible.

**Dependency:** None

**Notes:**

### PR DESCRIPTION
**Pull request by lgolding**
No fixer for this one. You can't safely make the type private; someone might be using it.

Note the special case handling for certain classes nested within `DataSet`. This requires us to add `System.Data` to the list of assemblies included within the generated test projects, since we don't have a way to customize the test projects at runtime.

@srivatsn @nguerrera @michaelcfanning @jinujoseph @basoundr @heejaechang

### CODE REVIEW

**Comment by srivatsn** [comment]
:+1:

File: src/Microsoft.ApiDesignGuidelines.Analyzers/Core/NestedTypesShouldNotBeVisible.cs [hunk 1]
```diff
@@ -19,31 +21,112 @@ public abstract class NestedTypesShouldNotBeVisibleAnalyzer : DiagnosticAnalyzer
...
+
+                            // The Framework Design Guidelines (see 4.9 Nested Types) say that it is okay
+                            // to expose nested types for advanced customization and subclassing scenarios,
+                            // so, following FxCop's implementation of this rule, we allow protected and
+                            // internal nested types.
+                            if (nestedType.DeclaredAccessibility != Accessibility.Public)
+                            {
+                                return;
+                            }
+
```
Comment (nguerrera) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/NestedTypesShouldNotBeVisible.cs]: I'm not seeing where we check that the type is visible from outside the assembly. 

FxCop would not warn on:

```
internal class Outer {
    public class Inner {}
}
```
This was implemented via the `TargetVisibility` override returning `ExternallyVisible`.

Comment (srivatsn) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/NestedTypesShouldNotBeVisible.cs]: Interesting. @lgolding  you can check that with `nestedType.GetResultantVisibility() == SymbolVisibility.Public`

Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/NestedTypesShouldNotBeVisible.cs]: Good catch. That is a bug. I'll add failing test cases and then fix it.

**Pull request merged**